### PR TITLE
Update input field readOnly and selectable

### PR DESCRIPTION
**Pull request by vinothvino42**
Ticket: [#1220](https://github.com/EnsembleUI/ensemble/issues/1220)

Acceptance criteria:
1. Default values are `readOnly: false` and `selectable: true`
2. When I only set `readOnly: true`, and do not set enabled or selectable props, user should be able to select the value and copy it. User should not be able to modify the value.
3. When I only set `enabled: fase`, widget should ignore the `readOnly` prop. The widget will not be clickable at all.
4. When I set `readOnly` at the parent Form widget, the value must cascade to children of the Form that are TextInputs and PasswordInputs.

Sample EDL:

```yaml
View:
  header:
    title: Form
  styles:
    scrollableView: true

  body:
    Column:
      styles:
        gap: 16
        padding: 24
      children:
        - Form:
            styles:
              readOnly: true
            children:
              - TextInput:
                  id: name
                  label: Name
                  value: EnsembleUI
                  required: true
                  readOnly: false
                  # enabled: false

              - PasswordInput:
                  label: Password
                  value: ABCD
                  obscureToggle: true
                  # readOnly: false
                  # enabled: false

```

### CODE REVIEW

File: lib/widget/input/form_textfield.dart [hunk 1]
```diff
@@ -217,7 +216,7 @@ class TextInputController extends FormFieldController {
 
   // applicable only for Password or obscure TextInput, to toggle between plain and secure text
   bool? obscureToggle;
-  bool readOnly = false;
+  bool? readOnly;
   bool selectable = true;
```
Comment (vusters) [inline lib/widget/input/form_textfield.dart:220]: Why is this still here?

File: lib/layout/form.dart [hunk 2]
```diff
@@ -158,6 +161,15 @@ class FormState extends WidgetState<EnsembleForm>
     for (Widget formItem in formItems) {
       if (formItem is HasController &&
           formItem.controller is FormFieldController) {
+        // Adding Form's readOnly value to the TextInput and PasswordInput
```
Comment (vusters) [inline lib/layout/form.dart:164]: This is wrong @vinothvino42 @snehmehta. The Form should have zero idea what a TextInput is, nor trying to reach inside its children and modify the child's data.

Comment (vusters) [reply lib/layout/form.dart:164]: Instead inside the child (i.g. TextInput) it should reach out to its parent Form and look at its attribute instead.

**Review by vusters — CHANGES_REQUESTED** [review]

File: lib/widget/input/form_textfield.dart (hunk 1 above)
Comment (vinothvino42) [reply lib/widget/input/form_textfield.dart:219]: Actually, both **TextInput** and **PasswordInput** field's controller extends from (`TextInputController -> FormFieldController -> WidgetController`)
But the Form widget's controller extends from (`FormController -> WidgetController`)
So added the `readOnly` property in two places (`TextInputController` and `FormController`)

`Form's controller's` `readOnly` property is to set all the `TextInput` and `PasswordInput` field's `readOnly` property in the **Form** widget

TextInput and PasswordInput's controller's readOnly property is used for `inline styling`, to `override` the parent widget (Form controller's readOnly property)

File: lib/layout/form.dart (hunk 2 above)
Comment (vinothvino42) [reply lib/layout/form.dart]: Refactored it. Now I accessed the parent Form's readOnly property from the child widget (BaseTextInput's state class) and assigned the value to it if the BaseTextInput controller's readOnly property is null

**Comment by vinothvino42** [comment]
@vusters Can you review it again?

File: lib/widget/input/form_textfield.dart [hunk 3]
```diff
@@ -317,6 +316,14 @@ class TextInputState extends FormFieldWidgetState<BaseTextInput>
         }
       }
     });
+
+    // Checking for readOnly from parent widget and assign the value to TextInput and PasswordInput if it's readOnly property is null
+    final formController =
```
Comment (vusters) [inline lib/widget/input/form_textfield.dart:321]: Dont' waste unnecessary cycles. There's no need to traverse the tree unless widget._controller.readOnly is not set.

File: lib/widget/input/form_textfield.dart [hunk 4]
```diff
@@ -317,6 +316,14 @@ class TextInputState extends FormFieldWidgetState<BaseTextInput>
...
+        context.findAncestorWidgetOfExactType<EnsembleForm>()?.controller;
+
+    if (formController != null && widget._controller.readOnly == null) {
+      widget._controller.readOnly = formController.readOnly == true;
```
Comment (vusters) [inline lib/widget/input/form_textfield.dart:325]: You are inheriting this property from Form, so it's not proper to overwrite it. This cause unnecessary bugs if the user is reading it back, and also waste another rendering cycle. You should simple do a getter `isReadOnly() => isReadOnly ?? form.isReadOnly`

**Review by vusters — CHANGES_REQUESTED** [review]
It's important to remove attributes you added inadvertently.

File: lib/widget/input/form_textfield.dart (hunk 1 above)
Comment (vinothvino42) [reply lib/widget/input/form_textfield.dart:219]: Why? `selectable` property also needed right. User can control (selectable: true or false) it anytime based on some condition

File: lib/widget/input/form_textfield.dart (hunk 4 above)
Comment (vinothvino42) [reply lib/widget/input/form_textfield.dart]: How can I get EnsembleForm widget's `formController` to get the readOnly property?

Our TextInput and PasswordInput field's controller extends from (`TextInputController -> FormFieldController -> WidgetController`)
But the Form widget's controller extends from (`FormController -> WidgetController`).

Now I want to get the `Form widget's readOnly` property and to set that parent value (Form's readOnly) to the children's (TextInput or PasswordInput) field's `readOnly` property

File: lib/widget/input/form_textfield.dart (hunk 3 above)
Comment (vinothvino42) [reply lib/widget/input/form_textfield.dart]: Updated it

**Review by vusters — APPROVED** [review]

**Pull request merged**